api/app/GraphQL/Mutations/UpdateUserSkillRankings.php

```diff
@@ -0,0 +1,95 @@
+<?php
+
+namespace App\GraphQL\Mutations;
+
+use App\Models\User;
+
+final class UpdateUserSkillRankings
+{
+    /**
+     * Update a user's UserSkill rankings
+     *
+     * @param  array{userId: UUID, userSkillRanking: UpdateUserSkillRankingsInput}  $args
+     * @return User
+     */
+    public function __invoke($_, array $args)
+    {
+        $user = User::find($args['userId'])->load(['userSkills', 'userSkills.skill']);
+        $userSkillsCollection = $user->userSkills;
+        $userSkillRankingInput = $args['userSkillRanking'];
+
+        // ensure UserSkill models exist for all input skills
+        $combinedSkillsId = array_merge(
+            $userSkillRankingInput['topTechnicalSkillsRanked'] ?? [],
+            $userSkillRankingInput['topBehaviouralSkillsRanked'] ?? [],
+            $userSkillRankingInput['improveTechnicalSkillsRanked'] ?? [],
+            $userSkillRankingInput['improveBehaviouralSkillsRanked'] ?? [],
+        );
+        $user->addSkills($combinedSkillsId);
+
+        // re-grab fresh User model with eager loaded new UserSkills
+        $user = User::find($args['userId'])->load(['userSkills', 'userSkills.skill']);
+        $userSkillsCollection = $user->userSkills;
+
+        // execute blocks depending on whether the value at args.userSkillRankingInput.X is non-null
+        if (isset($userSkillRankingInput['topTechnicalSkillsRanked'])) {
+
+            $this->syncRankings(
+                $userSkillsCollection,
+                $userSkillRankingInput['topTechnicalSkillsRanked'],
+                'TECHNICAL',
+                'top_skills_rank'
+            );
+        }
+
+        if (isset($userSkillRankingInput['topBehaviouralSkillsRanked'])) {
+
+            $this->syncRankings(
+                $userSkillsCollection,
+                $userSkillRankingInput['topBehaviouralSkillsRanked'],
+                'BEHAVIOURAL',
+                'top_skills_rank'
+            );
+        }
+
+        if (isset($userSkillRankingInput['improveTechnicalSkillsRanked'])) {
+
+            $this->syncRankings(
+                $userSkillsCollection,
+                $userSkillRankingInput['improveTechnicalSkillsRanked'],
+                'TECHNICAL',
+                'improve_skills_rank'
+            );
+        }
+
+        if (isset($userSkillRankingInput['improveBehaviouralSkillsRanked'])) {
+
+            $this->syncRankings(
+                $userSkillsCollection,
+                $userSkillRankingInput['improveBehaviouralSkillsRanked'],
+                'BEHAVIOURAL',
+                'improve_skills_rank'
+            );
+        }
+
+        $user->refresh();
+
+        return $user;
+    }
+
+    private function syncRankings(object $userSkillsCollection, array $arraySkillIds, string $skillCategory, string $rankType)
+    {
+        // clear existing ranking for the category and type passed in
+        $userSkillsCollection->where('skill.category', $skillCategory)
```

Comment:
Thinking about that `each` question again...
We could make this slightly more performant by doing it in one DB call by not using userSkillsCollection at all (meaning we could even skip loading them in the first place) and just doing `UserSkills::where('user_id', $userId)->whereHas('skill', function ($query) { $query->where('category', $skillCategory);})->update(...)`. Similarly for below, start with `UserSkills::where('user_id', ...)` instead of userSkillsCollection.

Reply:
Hm... Okay, sure. The collection was straightforward to remove, I don't know how much this changed performance
Done in https://github.com/GCTC-NTGC/gc-digital-talent/pull/7678/commits/e7425b61874691f1d8dae5820451beb401a42469

Reply:
Oh, didn't need a `whereHas` instance
https://github.com/GCTC-NTGC/gc-digital-talent/pull/7678/commits/bde556c404ee36ef33a2d36e61e209e42829b450